add rudimentary integration-style tests

aktau opened the pull request:
It uses sleeps,  but it's better than nothing, and due to using the public API (command-line arguments), no changes to perfleck were needed at all. The tests run in ~1.6s. I use the following to run test when I change anything:

    $ go install github.com/tmc/watcher
    $ watcher -v -depth 3 -dir ../.. go test . -count 10

If I would've had these when I started hacking on perflock, I would've saved myself some time.

aclements commented on cmd/perflock/daemon.go:
```diff
@@ -28,10 +28,12 @@ func doDaemon(path string) {
-	err = os.Chmod(path, 0777)
-	if err != nil {
-		log.Fatal(err)
+	if len(path) > 1 && path[0] != '@' {
```
This is pretty opaque. Leave a comment?

aktau replied:
Done.

aclements commented on cmd/perflock/perflock_test.go:
```diff
@@ -0,0 +1,214 @@
+func TestExclusive(t *testing.T) {
+	t.Parallel()
+
+	socket := fmt.Sprintf("@perflock.%d.%s", os.Getpid(), funcname())
```
I'm not sure you really need the abstract socket. Couldn't this just be a regular UNIX domain socket in a directory created by t.TempDir()? That would also avoid the need for the PID and funcname()

aktau replied:
It could be, but then tests leave garbage around. Abstract domain sockets automatically clean themselves up, don't hang when there's nothing on the other side and can't have any problems if whoever is running the tests has a really messed up filesystem (such as an unwriteable temp). It just avoids some gotchas.

aktau replied:
At this point I've:

 1. Split the support for abstract domain sockets into a separate commit.
 2. Use them in tests on Linux. This serves as a way of testing this support as well. Tests should also work on macOS, as on non-Linux OSen it uses socket files.

aclements commented on cmd/perflock/perflock_test.go:
```diff
@@ -0,0 +1,214 @@
+	pid = cmd.Process.Pid
+	t.Cleanup(func() {
+		cancel()
+		if err := cmd.Wait(); err != nil {
```
Just for robustness, set cmd.WaitDelay?

aktau replied:
I'm developing this on a Debian stable (Crostini on a Chromebook), so the version is 1.19 and WaitDelay was introduced in 1.20. I've now upgraded my toolchain to 1.21 so I can use the latest and greatest stuff. I've updated the go.mod file for the new minimum version.

aclements commented on cmd/perflock/perflock_test.go:
```diff
@@ -0,0 +1,214 @@
+	go func() {
+		envs := strings.Join(env, " ")
+		for scanner.Scan() {
+			fmt.Printf("[%12d] %-25s %s\n", pid, envs, scanner.Text())
```
t.Logf?

aktau replied:
Good call. Was a holdover from when I was subprocessing too much.

aktau commented:
I've updated a new version (force push) addressing your comments, except for the abstract domain sockets, which I still think are superior to filesystem names.

aktau committed:
support the abstract namespace for UDS sockets on Linux

This avoids filesystem operations, see unix(7). Could be useful for
people who have a fully read-only filesystem. Also no need to chmod(1)
to make the socket connectable by anyone (that's the default).

aktau committed:
add rudimentary integration-style tests

It uses sleeps, but it's better than nothing. Due to using the public
API (command-line arguments), no changes to perfleck were needed. The
tests run in ~1.6s. I use the following to run test when I change
anything:

    $ go install github.com/tmc/watcher
    $ watcher -v -depth 3 -dir ../.. go test . -count 10

If I would've had these when I started hacking on perflock, I would've
saved myself some time.

Minimum Go version is increased from 1.15 to 1.20 due to use of
`os/exec.Command.WaitDelay`.

aclements commented on cmd/perflock/perflock_test.go:
```diff
@@ -0,0 +1,229 @@
+
+// socketName returns a unique socket name per test.
+func socketName(t *testing.T) string {
+	if runtime.GOOS == "linux" && false {
```
I see the `&& false`. Should this path be deleted? Or revived?

aktau replied:
Sorry, that was a leftover from when I was testing that both paths still work on Linux (meaning I could be more sure that it would work on Darwin). Removed it now.
